tooling/debugger/src/context.rs

```diff
@@ -15,56 +16,138 @@ use nargo::errors::{ExecutionError, Location};
 use nargo::NargoError;
 use noirc_driver::DebugFile;
 
+use thiserror::Error;
+
 use std::collections::BTreeMap;
 use std::collections::{hash_set::Iter, HashSet};
 
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub struct DebugLocation {
+    pub circuit_id: u32,
+    pub opcode_location: OpcodeLocation,
+}
+
+impl std::fmt::Display for DebugLocation {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let circuit_id = self.circuit_id;
+        match self.opcode_location {
+            OpcodeLocation::Acir(index) => write!(f, "{circuit_id}:{index}"),
+            OpcodeLocation::Brillig { acir_index, brillig_index } => {
+                write!(f, "{circuit_id}:{acir_index}.{brillig_index}")
+            }
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum DebugLocationFromStrError {
+    #[error("Invalid debug location string: {0}")]
+    InvalidDebugLocationString(String),
+}
+
+impl std::str::FromStr for DebugLocation {
+    type Err = DebugLocationFromStrError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts: Vec<_> = s.split(':').collect();
+
+        if parts.is_empty() || parts.len() > 2 {
+            return Err(DebugLocationFromStrError::InvalidDebugLocationString(s.to_string()));
+        }
+
+        fn parse_components(parts: &Vec<&str>) -> Option<DebugLocation> {
+            match parts.len() {
+                1 => {
+                    let opcode_location = OpcodeLocation::from_str(parts[0]).ok()?;
+                    Some(DebugLocation { circuit_id: 0, opcode_location })
+                }
+                2 => {
+                    let circuit_id = parts[0].parse().ok()?;
+                    let opcode_location = OpcodeLocation::from_str(parts[1]).ok()?;
+                    Some(DebugLocation { circuit_id, opcode_location })
+                }
+                _ => unreachable!("`DebugLocation` has too many components"),
+            }
+        }
+
+        parse_components(&parts)
+            .ok_or(DebugLocationFromStrError::InvalidDebugLocationString(s.to_string()))
+    }
+}
+
 #[derive(Debug)]
 pub(super) enum DebugCommandResult {
     Done,
     Ok,
-    BreakpointReached(OpcodeLocation),
+    BreakpointReached(DebugLocation),
     Error(NargoError),
 }
 
+pub struct ExecutionFrame<'a, B: BlackBoxFunctionSolver<FieldElement>> {
+    circuit_id: u32,
+    acvm: ACVM<'a, FieldElement, B>,
```

Comment:
I'm wary that need to store the entire ACVM here? I only saw it really being used to fetch the instruction pointer and status. If there is something I missed could you elaborate?

Reply:
This is used to suspend the execution of an ACIR circuit while resolving an ACIR call, similar to how it's done [here](https://github.com/noir-lang/noir/blob/master/tooling/nargo/src/ops/execute.rs#L67). Only we cannot use recursion and the stack in this case, so we need to improvise our own stack.

An alternative would be to serialize the ACVM state and after the ACIR call is resolved, restore the parent ACVM with the serialized state. But that would require changes in the ACVM code itself.

Reply:
Ah yeah I understand. I don't think we should change the ACVM interfaces for now and de/serializing the entire ACVM state is probably more overhead than just storing the ACVM stack in the way you have here. We can leave this as is.